bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/composite/AbstractCompositeModuleHandler.java

```diff
@@ -88,10 +89,14 @@ public AbstractCompositeModuleHandler(M module, MT moduleType, LinkedHashMap<M,
 
     @Override
     public void dispose() {
-        if (moduleHandlerMap != null) {
-            moduleHandlerMap.clear();
-            moduleHandlerMap = null;
+        List<M> children = getChildren();
+        for (M child : children) {
+            ModuleHandler childHandler = moduleHandlerMap.remove(child);
+            childHandler.dispose();
         }
```

Comment:
Yes, you have a point, there is a bit of similarity in both mechanisms, but :

1. In DS we have the bind/unbind methods - where  implemenetator for sure knows what to do in the unbind 
In our case we have just the setRuleEnginceCallback, and imho, calling one and the same method , once with real reference, once with null, will only harden the implementation - having always logic isNull then do that, so that there are no unexpected NPE's in some clients handlers where:
1.1 would want to trigger the rule with some initial values, right after the ruleCallback is atached
1.2 would expect setRuleCallback to be called only once with the real reference and in that method to initalize for example an eventHandler - which would again cause a NPE if first dispose is called, where you have released some of the needed references).
 

If we do agree that dispose is not enough I think we have 2 options:
- Adding unsetRuleEngineCallback method in the TriggerHandler interface - matching the mechanism of bind/unbind.
- Propose BaseTriggerHandler class that in its dispose method would just releases the reference of ruleCallback 


But for sure I do not think its a good idea to call setRuleEngineCallback => null, when having another method(dispose) that actually its purpose is to release all used references, because this would work only if all TriggerHandlers: 
1. expect null value and somewhere check if ruleCallBack isNotNull
2. using the setRuleEngineCallback only for initalizing the ruleCallback itself (which is similar to 1.)

@danchom What do you think?